conbench/entities/benchmark_result.py

```diff
@@ -138,168 +142,116 @@ def create(userres):
-        # business logic treat it as such.)
-        if result_error_for_db is not None:
-            benchmark_result_data["error"] = result_error_for_db
-
-        # Temporary: keep name `data` -- it's unfortunate that we have the name
-        # `data` here while also having key(s) in the dict(s) that are called
-        # `data`. Assign a recognizable name to the big, outest, user-given
-        # object: userres is not ideal, but a start.
-        data = userres
+        else:
+            # process_samples_build_agg() must only be called if
+            # do_iteration_samples_look_like_error() returned False. That's
+            # the case here.
+            result_data_from_stats = validate_and_aggregate_samples(userres["stats"])
+
+            # Per-iteration samples looked good, and we did (potentially)
+            # rebuild aggregates. Merge dict `result_stats_data_for_db` on top
+            # of dict `benchmark_result_data`, overwriting upon conflict.
+            result_data_for_db |= result_data_from_stats  # PEP 584 update
 
+        # See https://github.com/conbench/conbench/issues/935,
         # At this point, assume that data["tags"] is a flat dictionary with
         # keys being non-empty strings, and values being non-empty strings.
-        tags = data["tags"]
-        has_error = "error" in data
-
-        if "stats" in data:
-            # Note(JP): this is the result of marshmallow
-            # deserialization/validation against the `BenchmarkResultCreate`
-            # schema below. This needs better naming. Most importantly, we
-            # should make use of typing information derived from the schema.
-
-            # Merge this on top of the above-defined benchmark_result_data.
-            benchmark_result_data = benchmark_result_data | data["stats"].copy()
-
-            if benchmark_result_data.get("data"):
-                # Note(JP): looks like `benchmark_result_data["data"]` is a
-                # list of either `None` or `Decimal` objects (currently
-                # enforced via marshmallow schema). Example:
-                #
-                # [Decimal('0.099094'), None, Decimal('0.036381')]
-                #
-                # TODO: let's lock this in with type annotations so that this
-                # is known _here_. Transform `None` to `math.nan` and `Decimal``
-                # to `float`` for subsequent calculations.
-                dat = [
-                    float(x) if x is not None else math.nan
-                    for x in benchmark_result_data["data"]
-                ]
-
-                q1 = q3 = mean = median = min = max = stdev = iqr = None
-
-                # calculate stats only if the data is complete. Note(JP): why?
-                # maybe we want to calculate that always if there are at least
-                # N (3?) data points.
-                any_errored_iteration = any([x is math.nan for x in dat])
-
-                if not any_errored_iteration:
```

Comment:
When code grows far to the bottom right it gets harder and harder to read.

So, this entire code section is now split into some units of work that are maybe a little easier to handle, and so that the very high-level business logic is easier to infer.